From c0bd5f75c90ebd074e386586f4d9d4556591ce9b Mon Sep 17 00:00:00 2001
From: Daniel Johnson <hexahedria@gmail.com>
Date: Thu, 19 Nov 2015 20:04:11 -0800
Subject: [PATCH 8/8] Change dumping to check for current mode spsr

The spsr for the current mode is stored separately from the banked versions
for other modes. Before, this meant when we dumped state we pulled down an
out-of-date version of the spsr. Update to check what mode we are in and
use the current spsr if we are dumping for the current mode.
---
 target-arm/translate.c |   12 +++++++++++-
 1 files changed, 11 insertions(+), 1 deletions(-)

diff --git a/target-arm/translate.c b/target-arm/translate.c
index 61ce88e..db53ca3 100644
--- a/target-arm/translate.c
+++ b/target-arm/translate.c
@@ -11556,7 +11556,17 @@ unsigned int get_register_in_mode(CPUARMState *env, int mode, int i) {
     return env->regs[i];
 }
 unsigned int get_spsr_in_mode(CPUARMState *env, int mode) {
-    return env->banked_spsr[bank_number(mode)];
+    int cur_mode;
+
+    cur_mode = env->uncached_cpsr & CPSR_M;
+    if (mode == -1)
+        mode = cur_mode;
+    
+    if (mode == cur_mode) {
+        return env->spsr;
+    } else {
+        return env->banked_spsr[bank_number(mode)];
+    }
 }
 
 unsigned int get_coprocessor_reg(ARMCPU *cpu, int crn) {
-- 
1.7.1

